lxd/storage/drivers/driver_powerflex_utils.go

```diff
@@ -0,0 +1,1021 @@
+func (p *powerFlexClient) login() error {
+	if p.token.Get() != "" {
+		return nil
+	}
+
+	body, err := p.createBodyReader(map[string]any{
+		"username": p.driver.config["powerflex.user.name"],
+		"password": p.driver.config["powerflex.user.password"],
+	})
+	if err != nil {
+		return err
+	}
+
+	var actualResponse struct {
+		AccessToken string        `json:"access_token"`
+		ExpiresIn   time.Duration `json:"expires_in"`
+	}
+
+	err = p.request(http.MethodPost, "/rest/auth/login", body, &actualResponse)
+	if err != nil {
+		return fmt.Errorf("Failed to login: %w", err)
+	}
+
+	// Expire the token after half of its lifetime
+	p.token.Set(actualResponse.AccessToken, actualResponse.ExpiresIn*time.Second/2)
+	return nil
+}
+
+// getStoragePool returns the storage pool behind poolID.
+func (p *powerFlexClient) getStoragePool(poolID string) (*powerFlexStoragePool, error) {
+	var actualResponse powerFlexStoragePool
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/StoragePool::%s", poolID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get storage pool: %q: %w", poolID, err)
+	}
+
+	return &actualResponse, nil
+}
+
+// getProtectionDomain returns the protection domain behind domainID.
+func (p *powerFlexClient) getProtectionDomain(domainID string) (*powerFlexProtectionDomain, error) {
+	var actualResponse powerFlexProtectionDomain
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/ProtectionDomain::%s", domainID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get protection domain: %q: %w", domainID, err)
+	}
+
+	return &actualResponse, nil
+}
+
+// getProtectionDomainSdsRelations returns the protection domains SDS relations.
+func (p *powerFlexClient) getProtectionDomainSdsRelations(domainID string) ([]powerFlexProtectionDomainSdsRelation, error) {
+	var actualResponse []powerFlexProtectionDomainSdsRelation
+	err := p.requestAuthenticated(http.MethodGet, fmt.Sprintf("/api/instances/ProtectionDomain::%s/relationships/Sds", domainID), nil, &actualResponse)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get protection domain SDS relations: %q: %w", domainID, err)
+	}
+
+	return actualResponse, nil
+}
+
+// getVolumeID returns the volume ID for the given name.
+func (p *powerFlexClient) getVolumeID(name string) (string, error) {
+	body, err := p.createBodyReader(map[string]any{
+		"name": name,
+	})
+	if err != nil {
+		return "", err
+	}
+
+	var actualResponse string
+	err = p.requestAuthenticated(http.MethodPost, "/api/types/Volume/instances/action/queryIdByKey", body, &actualResponse)
+	if err != nil {
+		powerFlexError, ok := err.(*powerFlexError)
+		if ok {
+			// API returns 500 if the volume does not exist.
+			// To not confuse it with other 500 that might occur check the error code too.
+			if powerFlexError.HTTPStatusCode() == 500 && powerFlexError.ErrorCode() == 79 {
```

Comment:
We can use http.StatusInternalServerError here

Also, can we define 79 (and other powerflex error codes) as constants rather than having magic numbers in logic themselves.

Reply:
Moved to a constant and using the `http` packages ones for comparison.